example/example-knex-mysql.js

```diff
@@ -0,0 +1,141 @@
+// https://blog.nodeswat.com/concurrency-mysql-and-node-js-a-journey-of-discovery-31281e53572e
+//
+// Below is an example table to store rate limits that must be created
+// in the database first.
+//
+// exports.up = async knex => {
+//   await knex.schema.createTable('rate_limits', table => {
+//     table.string('source').notNullable()
+//     table.string('route').notNullable()
+//     table.integer('count').unsigned()
+//     table.bigInteger ('ttl')
+//     table.primary(['route', 'source'])
+//   })
+// }
+//
+// exports.down = async knex => {
+//   await knex.schema.dropTable('rate_limits')
+// }
+//
+// CREATE TABLE `rate_limits` (
+//   `source` varchar(255) NOT NULL,
+//   `route` varchar(255) NOT NULL,
+//   `count` int unsigned DEFAULT NULL,
+//   `ttl` int unsigned DEFAULT NULL,
+//   PRIMARY KEY (`route`,`source`)
+// ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
+
+function KnexStore(options) {
+  this.options = options
+  this.route = ''
+}
+
+KnexStore.prototype.routeKey = function (route) {
+  if (route) this.route = route
+  return route
+}
+
+KnexStore.prototype.incr = async function (key, cb) {
+  const now = (new Date()).getTime()
+  const ttl = now + this.options.timeWindow
+  const max = this.options.max
+  const cond = { route: this.route, source: key }
+  await knex.transaction(async (trx) => {
+    try {
+      // NOTE: MySQL syntax FOR UPDATE for read lock on counter stats in row
+      const row = await trx('rate_limits')
+        .whereRaw('route = ? AND source = ? FOR UPDATE', [cond.route || '', cond.source]) // Create read lock
+      const d = row[0]
+      if (d && d.ttl > now) {
+        // Optimization - no need to UPDATE if max has been reached.
+        if(d.count < max) {
+          try {
+            await trx
+              .raw('UPDATE rate_limits SET count = ? WHERE route = ? AND source = ?', [d.count + 1, cond.route, key])
+            cb(null, { current: d.count + 1, ttl: d.ttl })
+          } catch(err) {
+            // TODO: Handle as appropriate
+            fastify.log.error(err)
+            cb(err, { current: 0 })
+          }
+        } else { // We're already at max. No UPDATE above saves a write, but we must send d.count + 1 to trigger rate limit.
+          cb(null, { current: d.count +1, ttl: d.ttl })
+        }
+      } else {
+        try {
+          // NOTE: MySQL syntax for ON DUPLICATE KEY UPDATE
+          await trx
+            .raw('INSERT INTO rate_limits(route, source, count, ttl) VALUES(?,?,1,?) ON DUPLICATE KEY UPDATE count = 1, ttl = ?', [cond.route, key, (d && d.ttl) || ttl, ttl])
+          cb(null, { current: 1, ttl: (d && d.ttl) || ttl })
+        } catch(err) {
+          // TODO: Handle as appropriate
+          fastify.log.error(err)
+          cb(err, { current: 0 })
+        }
+      }
+    } catch(err) {
+      // TODO: Handle as appropriate
+      fastify.log.error(err)
+      cb(err, { current: 0 })
```

Comment:
```suggestion
      process.nextTick(cb, err, { current: 0 })
```

This fixes the error handling of using async/await with a callback.

Reply:
Thanks

Reply:
I've wrapped each callback in process.nextTick as suggested.